File: 프로그래머스/2/49994. 방문 길이/방문 길이.js
function solution(dirs) {
    let x = 0
    let y = 0
    const directions = {
        'U' : [0, 1],
        "D" : [0, -1],
        "R" : [1, 0],
        "L" : [-1, 0]
    }
    
    let path = {}
    let res = 0
    for(let dir of dirs){
        let [dx, dy] = directions[dir]
        let cx = x + dx
        let cy = y + dy
        
        if(cx >= -5 && cx <= 5 && cy >= -5 && cy <= 5){
            let key1 = `${x}${y}${cx}${cy}`
            let key2 = `${cx}${cy}${x}${y}`
            if(!path[key1] && !path[key2]){
                path[key1] = true
                path[key2] = true
                res++
            }

            x = cx
            y = cy
        }
    }
    
    return res
}

File: 프로그래머스/2/49994. 방문 길이/방문 길이.java
import java.util.*;

class Solution {
    public int solution(String dirs) {
        int x = 0;
        int y = 0;
        
        Set<String> visited = new HashSet<>();
        
        int cnt = 0;
        
        for(char dir : dirs.toCharArray()){
            int nx = x;
            int ny = y;
            
            if(dir == 'U') ny++;
            else if(dir == 'D') ny--;
            else if(dir == 'R') nx++;
            else if(dir == 'L') nx--;
            
            if(nx < -5 || nx > 5 || ny < -5 || ny > 5) continue;
            
            String key1 = x + "," + y + ":" + nx + "," + ny;
            String key2 = nx + "," + ny + ":" + x + "," + y;
            if(!visited.contains(key1)){
                visited.add(key1);
                visited.add(key2);
                cnt++;
            }
            
            x = nx;
            y = ny;
        }
        
        return cnt;
    }
}